src/map/core/raw.rs

```diff
@@ -0,0 +1,276 @@
+    where
+        Q: ?Sized + Equivalent<K>,
+    {
+        self.indices.find(hash.get(), {
+            move |&i| Q::equivalent(key, &self.entries[i].key)
+        })
+    }
+
+    /// Return the raw bucket for the given index
+    fn find_index(&self, hash: HashValue, index: usize) -> Option<RawBucket> {
+        self.indices.find(hash.get(), move |&i| i == index)
+    }
+
+    /// Return the index in `entries` where an equivalent key can be found
+    pub(crate) fn get_index_of<Q>(&self, hash: HashValue, key: &Q) -> Option<usize>
+    where
+        Q: ?Sized + Equivalent<K>,
+    {
+        match self.find_equivalent(hash, key) {
+            Some(raw_bucket) => Some(unsafe { raw_bucket.read() }),
+            None => None,
+        }
+    }
+
+    pub(super) fn erase_index(&mut self, hash: HashValue, index: usize) {
+        let raw_bucket = self.find_index(hash, index).unwrap();
+        unsafe { self.indices.erase_no_drop(&raw_bucket) };
+    }
+
+    pub(crate) fn entry(&mut self, hash: HashValue, key: K) -> Entry<K, V>
+    where
+        K: Eq,
+    {
+        match self.find_equivalent(hash, &key) {
+            // Safety: The entry is created with a live raw bucket, at the same time we have a &mut
+            // reference to the map, so it can not be modified further.
+            Some(raw_bucket) => Entry::Occupied(OccupiedEntry {
+                map: self,
+                raw_bucket,
+                key,
+            }),
+            None => Entry::Vacant(VacantEntry {
+                map: self,
+                hash,
+                key,
+            }),
+        }
+    }
+
+    /// Remove an entry by shifting all entries that follow it
+    pub(crate) fn shift_remove_full<Q>(&mut self, hash: HashValue, key: &Q) -> Option<(usize, K, V)>
+    where
+        Q: ?Sized + Equivalent<K>,
+    {
+        match self.find_equivalent(hash, key) {
+            Some(raw_bucket) => unsafe { Some(self.shift_remove_bucket(raw_bucket)) },
+            None => None,
+        }
+    }
+
+    /// Remove an entry by shifting all entries that follow it
+    pub(crate) fn shift_remove_index(&mut self, index: usize) -> Option<(K, V)> {
+        let raw_bucket = match self.entries.get(index) {
+            Some(entry) => self.find_index(entry.hash, index).unwrap(),
+            None => return None,
+        };
+        unsafe {
+            let (_, key, value) = self.shift_remove_bucket(raw_bucket);
+            Some((key, value))
+        }
+    }
+
+    /// Remove an entry by shifting all entries that follow it
+    ///
+    /// Safety: The caller must pass a live `raw_bucket`.
+    #[allow(unused_unsafe)]
+    unsafe fn shift_remove_bucket(&mut self, raw_bucket: RawBucket) -> (usize, K, V) {
+        // use Vec::remove, but then we need to update the indices that point
+        // to all of the other entries that have to move
+        let index = unsafe {
+            self.indices.erase_no_drop(&raw_bucket);
+            raw_bucket.read()
```

Comment:
I don't think this matters are the moment, but you should probably do the `read` before you erase the bucket, since _technically_ the map can do whatever it wants to the bucket once you've erased it. I ran into this when trying to use [`griddle`] as a drop-in replacement from `hashbrown`, since it _does_ occasionally invalidate the bucket's memory when it is erased.

Reply:
Hmm, `hashbrown` uses this pattern itself a lot, e.g. in `OccupiedEntry`:

```rust
    pub fn remove_entry(self) -> (K, V) {
        unsafe {
            self.table.table.erase_no_drop(&self.elem);
            self.elem.read()
        }
    }
```

@Amanieu, can you clarify if this is something we can rely on? We should probably update the documentation of `erase_no_drop` either way.

> since it _does_ occasionally invalidate the bucket's memory when it is erased.

I can't see how that would happen, unless you allow other mutating operations before you actually read it.

Not that it's a big deal to reorder this, just a little more awkward.

Reply:
So, in `griddle` the reason this can happen is that `erase_no_drop` may erase the last non-empty bucket from a table that we were migrating items away from. If that is the case, then that is the best time for griddle to de-allocate it. It _could_ realize this later if we strictly need the contract for `erase_no_drop` to be this way, but it seems nicer if `erase_no_drop` isn't _required_ to continue to leave the bucket available after it is called.

Reply:
`erase_no_drop` only marks the bucket as deleted in the control bytes. It doesn't touch the actual data and it is fine to read from it. In fact hashbrown relies on this when dropping values in-place since `drop_in_place` may panic but we still want to ensure the bucket is marked as deleted if that happens.

Reply:
I guess my argument here is that while you _can_ rely on it for `hashbrown`, it'd be nice if that wasn't a part of `erase_no_drop`'s _contract_, since it means that anyone who wants to mimick must also uphold that contract. Or alternatively, it'd be great if `indexmap` did not rely on this aspect of `erase_no_drop` (though admittedly for selfish reasons). It's of course fine for `hashbrown` to continue to depend on this internally, since it's not going to use anyone else's `raw` any time soon :p

In griddle, I got around `drop_in_place` by instead doing a `read` + `erase_no_drop` + `drop`. It's not _quite_ as efficient, but it does mean that `erase_no_drop` is free to reclaim memory immediately should it so wish.

Reply:
For what it's worth, I also agree with @cuviper that if this is indeed something that `erase_no_drop` guarantees, then it should be documented that that is the case, especially since the name does not suggest that that guarantee exists. If `erase_no_drop` is updated that way, then I'll also update `griddle` to match, even if it adds some reclamation overhead.

Reply:
I filed this as https://github.com/rust-lang/hashbrown/issues/168. Depending on the resolution to that, this code should either stay the way it is (if it becomes a part of `erase_no_drop`'s guarantees), or the order of the operations should be reversed (if it does not).

Reply:
> `erase_no_drop` may erase the last non-empty bucket from a table that we were migrating items away from. If that is the case, then that is the best time for griddle to de-allocate it.

This feels a little contradictory. I guess you can narrowly interpret "no drop" to refer to the item only, but dropping the allocation is still a relatively heavy operation for what the method name implies.

Reply:
Yeah, I suppose that's true. I almost wonder if `erase_no_drop` should be entirely replace with something like `erase_and_take`? Or maybe `erase_and_take` + `erase_and_drop`.

Reply:
Now that https://github.com/rust-lang/hashbrown/pull/171 has landed, I think this should probably be merged as-is (to avoid holding it up), and then we should move to `erase`/`remove` once `hashbrown 0.8.1` is out.